### ISSUE TITLE
False positives typechecking function types

### ISSUE THREAD

**Issue by jaqmol** [microsoft/TypeScript#36770]
I searched for already reported bugs concerning this issue, but couldn't find any. In case some kind of explanation for the observed buggy behavior exists, I must excuse me: As to the nature of reality a myriad if different formulations for the same thing is actually possible and unfortunately I'm incapable to cover all of them.

I'm trying to implement a simple when() function for use as an conditional expression.

**Code**

```ts
type WhenThenFn<ThenT> = () => ThenT;
type WhenElseFn<ElseT> = () => ElseT;

function when<ThenT, ElseT>(
  condition: boolean|(() => boolean),
  thenArg: ThenT|WhenThenFn<ThenT>,
  elseArg: ElseT|WhenElseFn<ElseT>,
) :(ThenT|ElseT) {
  const condResult = typeof condition === 'function' ? condition() : condition;
  if (condResult)
    if (typeof thenArg === 'function') {
      const thenFn = thenArg as (() => ThenT);
      return thenFn();
    } else return thenArg;
  }
  if (typeof elseArg === 'function') {
    const elseFn = elseArg as (() => ElseT);
    return elseFn();
  } else return elseArg;
}
```

**Expected behavior:**

Just works.

**Actual behavior:**

#### Line `) :(ThenT|ElseT) {` => Function lacks ending return statement and return type does not include 'undefined'.ts(2366)

This statement is definitely false.

#### Line `const elseFn = elseArg as WhenElseFn<ElseT>;` => Cannot find name 'ElseT'.ts(2304)

Concerning this statements I'm seriously suspecting that the current implementation of the TypeScript typechecker is broken and should be reconsidered.

**[TypeScript Playground Link](http://www.typescriptlang.org/play/?ssl=2&ssc=22&pln=1&pc=1#code/C4TwDgpgBA6gFhAdgFQYgYogPKpyB8UAvFABQCUxhuKA3AFCiSxoCiANgM4SZYfcFiZSkUL8IyBvQBmAV0QBjYAEsA9oigB3NDjTIANFHEFS9KFAXqAJspXqAXFABGq1ewgBDRAB9SFKs6u7l7k+mZQwGgAggBOAOaONMje8HhovEn4YeYQXBCxCUZ5yamI4rzGWfSU9qRJ3saUAN7hloicwBbWAEoQnLLsnSRMEKrSXYg2dhpEs1AA5HKK0-NQAPwTU2qI-o5tW+oM5srjpPu9-YPk4cenI2MR0fHEc4vyStvzzTfmEx2PSEwQkiSAKUA8nBYaUB2EyR1+5hiEGAshiGhBGB25Hh5gAvlBctwoEiUWiAYgCvDceETmR7uNCflnrMSG9lp9vgi2v9GUCSIywRCoWU8hVivgccTkaiNLysVSCXkpaTZXlKfRcUA)**

#### PS: Here an implementation that is fine for the typechecker

```ts
type WhenFn<T> = () => T;

function when<ThenT, ElseT>(
  condArg: boolean|WhenFn<boolean>,
  thenArg: ThenT|WhenFn<ThenT>,
  elseArg: ElseT|WhenFn<ElseT>,
) :(ThenT|ElseT) {
  function ex<T>(arg: T|WhenFn<T>) :T {
    if (typeof arg === 'function') {
      const argFn = arg as WhenFn<T>;
      return argFn();
    } else return arg as T;
  }
  const condResult = ex<boolean>(condArg);
  if (condResult) {
    return ex<ThenT>(thenArg);
  } else {
    return ex<ElseT>(elseArg);
  }
}
```

**Comment by IllusionMH** [comment]
@jaqmol your playground example has missing `{` in line ` if (condResult)` therefore `if (typeof elseArg === 'function') {` and all code below is outside of a function where `elseArg` and `ElseT` are defined.
[Fixed playground](https://www.typescriptlang.org/play/#code/C4TwDgpgBA6gFhAdgFQYgYogPKpyB8UAvFABQCUxhuKA3AFCiSxoCiANgM4SZYfcFiZSkUL8IyBvQBmAV0QBjYAEsA9oigB3NDjTIANFHEFS9KFAXqAJspXqAXFABGq1ewgBDRAB9SFKs6u7l7k+mZQwGgAggBOAOaONMje8HhovEn4YeYQXBCxCUZ5yamI4rzGWfSU9qRJ3saUAN7hloicwBbWAEoQnLLsnSRMEKrSXYg2dhpEs1AA5HKK0-NQAPwTU2qI-o5tW+oM5srjpPu9-YPNUAD0N+ruEXDKnFDsyojQEFPf4cenIzGTyQBWIc0W8iU23mzT+5gmHWBGBmSNBHlepRoGT0+CO8PMMQgwFkMQ0kSQmAoePMAF8oLluFBCcTSaj4niaeETmRAeMGfl4mCSBDltDYfi2oj+ZghPy0Ri2HkKsVcXDmSSNNKduQOfS8kyiRq9dwCgwaUA)
